Source/WebGPU/WGSL/Metal/MetalFunctionWriter.cpp

```diff
@@ -92,16 +93,19 @@ class FunctionDefinitionWriter : public AST::Visitor {
 
 private:
     StringBuilder& m_stringBuilder;
-    ShaderModule& m_shaderModule;
+    CallGraph& m_callGraph;
     Indentation<4> m_indent { 0 };
     std::optional<AST::StructureRole> m_structRole;
     std::optional<AST::StageAttribute::Stage> m_entryPointStage;
     std::optional<String> m_suffix;
 };
 
-void FunctionDefinitionWriter::visit(ShaderModule& shaderModule)
+void FunctionDefinitionWriter::write()
 {
-    AST::Visitor::visit(shaderModule);
+    for (auto& structure : m_callGraph.ast().structures())
+        visit(structure);
```

Comment:
I assume you're going to prune these at some point too?

Reply:
That's correct, I still need to do that. I'm thinking the CallGraph should have a list of the used structs that we'd use instead of reading directly from the AST.